Tactics/Reflect/AxEffects.lean

```diff
@@ -0,0 +1,590 @@
+    eff.traceCurrentState
+    let fieldE ← mkFreshExprMVar (mkConst ``StateField)
+    let value ← mkFreshExprMVar none
+    let expectedType ← mkEq (mkApp2 (mkConst ``r) fieldE eff.initialState) value
+    assertHasType eq expectedType
+
+    let field ← reflectStateField (← instantiateMVars fieldE)
+    let fieldEff ← eff.getField field
+    trace[Tactic.sym] "current field effect: {fieldEff}"
+
+    if field ∉ eff.fields then
+      let proof ← mkEqTrans fieldEff.proof eq
+      let fields := eff.fields.insert field { value, proof }
+      return { eff with fields }
+    else
+      if eff.currentState == eff.initialState then
+        throwError "error: {field} already has an effect associated with it, \
+          and the current state
+            {eff.currentState}
+          is equal to the initial state
+            {eff.initialState}"
+
+      let valEq ← try rewrite fieldEff.value eq
+        catch _ => return eff
+
+      let_expr Eq _ _ value := ← inferType valEq
+        | throwError "internal error: expected an equality, found {valEq}"
+      let proof ← mkEqMP valEq fieldEff.proof
+      let fields := eff.fields.insert field { value, proof }
+      return { eff with fields }
+
+/-- Given a proof of `CheckSPAlignment <initialState>`,
+attempt to transport it to a proof of `CheckSPAlignment <currentState>`
+and store that proof in `stackAlignmentProof?`.
+
+Returns `none` if the proof failed to be transported,
+i.e., if SP was written to. -/
+def withStackAlignment? (eff : AxEffects) (spAlignment : Expr) :
+    MetaM (Option AxEffects) := do
+  let msg := m!"withInitialStackAlignment? {spAlignment}"
+  withTraceNode `Tactic.sym (fun _ => pure msg) <| do
+    eff.traceCurrentState
+
+    let { value, proof } ← eff.getField StateField.SP
+    let expected :=
+      mkApp2 (mkConst ``r) (toExpr <| StateField.SP) eff.initialState
+    trace[Tactic.sym] "checking whether value:\n  {value}\n\
+      is syntactically equal to expected value\n  {expected}"
+    if value != expected then
+      trace[Tactic.sym] "failed to transport proof:
+        expected value to be {expected}, but found {value}"
+      return none
+
+    let stackAlignmentProof? := some <|
+      mkAppN (mkConst ``CheckSPAlignment_of_r_sp_eq)
+        #[eff.initialState, eff.currentState, proof, spAlignment]
+    trace[Tactic.sym] "constructed stackAlignmentProof: {stackAlignmentProof?}"
+    return some { eff with stackAlignmentProof? }
+
+/-! ## Composition -/
+
+/- TODO: write a function that combines two effects `left` and `right`,
+where `left.initialState = right.currentState`.
+That is, compose the effect of "`left` after `right`" -/
+
+/-! ## Validation -/
+
+/-- type check all expressions stored in `eff`,
+throwing an error if one is not type-correct.
+
+NOTE: does not necessarily validate *which* type an expression has,
+validation will still pass if types are different to those we claim in the
+docstrings -/
+def validate (eff : AxEffects) : MetaM Unit := do
```

Comment:
A comment on why `validate` is run only under `sym.debug` would be nice. Performance reasons?

Also, another comment re. why `validate` is necessary in the first place. To debug the `Expr`s we create? Would the need for this function go away if we had a quoting mechanism?

Reply:
In a bug-free world, we only ever construct well-formed expressions, and running `validate` is a waste of time. Alas, I did write some bugs, so I wrote `validate` to help me catch exactly when I construct a bad expressions. 

So, a user would essentially never want to turn on `debug`, it's only there for us, the implementors of `sym_n`, to help in debugging